FIX: BDBD-494 구글 로그인 중복 회원가입 수정

joonamin opened the pull request:
### 개요
구글 로그인시, 기존 유저가 존재함에도 불구하고 중복 회원가입이 되었음
### 변경사항
loadUser시에 username 체크 로직 수정
### 관련 지라 및 위키 링크
`BDBD-494`
### 리뷰어에게 하고 싶은 말
* 항상 감사합니다 😊

joonamin committed:
FIX: 구글 로그인 중복 회원가입 수정

minseonglove commented on src/main/java/com/fakedevelopers/bidderbidder/service/OAuth2UserService.java:
```diff
@@ -75,7 +82,7 @@ public UserEntity loadUserOrRegister(String serviceProvider, FirebaseToken token
   }
 
-  /**
+  /**ㄴ
```
ㄴ

minseonglove commented on src/main/java/com/fakedevelopers/bidderbidder/service/OAuth2UserService.java:
```diff
@@ -65,7 +72,7 @@ public UserEntity loadUserOrRegister(String serviceProvider, FirebaseToken token
       OAuth2UserRegisterDto dto = OAuth2UserRegisterDto.builder()
           .username(
-              (username.substring(0, Math.min(token.getUid().length(), MAX_USERNAME_SIZE - 1))))
+              (username.substring(0, Math.min(serviceProvider.length() + token.getUid().length(), MAX_USERNAME_SIZE - 1))))
```
serviceProvider.length() + token.getUid().length()는 username.length()로 줄일 수 있겠네

minseonglove reviewed (approved):
잘햇서요

onikatashi commented on src/main/java/com/fakedevelopers/bidderbidder/service/OAuth2UserService.java:
```diff
@@ -43,8 +44,14 @@ public String getNaverClientSecret() {
+      throw new UsernameNotFoundException("User Not Found");
+    }
+    if (result.size() > 1) {
+      throw new AssertionError("[!FATAL] 동일 Username을 가진 유저가 2명 이상입니다");
```
동일한 username이 둘 이상 있을 때를 처리해줬네요! 좋아용

onikatashi reviewed (approved):
수고하셨습니다.

konempty commented on src/main/java/com/fakedevelopers/bidderbidder/repository/UserRepository.java:
```diff
@@ -13,7 +13,5 @@
-  Optional<UserEntity> findByEmail(String email);
-
-  Optional<UserEntity> findByUsername(String username);
+  List<UserEntity> findByUsername(String username);
```
오호..! 왜 리스트인지 궁금하구요 리스트를 받으려면 findAllBy로 받지않아도 되는걸까요?

joonamin replied:
[참고](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.query-methods.query-creation)기본 `findBy~~` 의 리턴값이 `List<>` 형태여서 맞추어준것인데, 추후에 좀 더 찾아보니
https://stackoverflow.com/questions/38147463/spring-data-findby-return-type
위와 같이 지원을 해주더라구요 ㅎㅎ! 추후에 테스트를 작성하기 위해서, 중복값이 있다면 해당 엔티티들을 모두 찾는다는 행위는 유지하고 싶은데 이 같은 경우에는 `findAllBy--` 가 좀 더 명시적일까요?? 의견이 궁금합니다

konempty replied:
@joonamin 아하 요건 저도 몰랐네요 ㅋㅋㅋ 요건 기존 코드중에 findAllBy를 쓰는 코드가 있으면 맞추는게 나을것 같고 없으면 그냥 지금도 괜찮아 보이네요!
그리고 추가로 궁금한게 있는데 유니크 해야 하는 값이면 table 제약 조건으로 unique를 걸고 select할때는 한개만 받아오도록 하는게 더 맞는것 같은데 어떻게 생각하시나요?

joonamin replied:
@konempty 
덕분에 현재 Transaction인 `register` 가 동일한 class 내부에서 호출되어 Roll-back이 안된다는 것을 확인했습니다 ㅎㅎ!
현재 table 제약 조건으로 unique를 걸었음에도 불구하고, commit이 되더라구요 이 부분은 proxy 객체 issue여서 수정한다음 commit 올리겠습니다.

konempty reviewed (approved):
굿굿 고생하셨습니다 👍 테스트가 잘 짜여져 있었다면 중복 회원가입이라는 불상사를 막을수 있지 않았을까 라는 생각이 쪼오금 드네요 ㅋㅋㅋㅋ..

joonamin committed:
chore: 리뷰 반영 -1

joonamin committed:
refactor: 리뷰 반영 - 1

joonamin committed:
test(`UserService`): UserService의 register()
* 정상적인 입력 값에 대해 commit이 성공적으로 수행되는 것을 확인

konempty reviewed (commented):
테스트 추가되면 말씀해주세요!

joonamin commented:
> 테스트 추가되면 말씀해주세요!

안녕하세요! 추가했습니다. 정당하지 않는 Input에 대한 테스트는 수행하질 못하였습니다. 알려주신다면 배우겠습니다 ㅎㅎ 감사합니다

joonamin committed:
chore: 코드 스멜 제거

joonamin committed:
fix: 테스트코드 수정
[원인](https://stackoverflow.com/questions/46850830/spring-jpa-doesnt-validate-bean-on-update/46851476)

konempty reviewed (approved):
고생하셨습니다 👍
